### DISCUSSION TITLE
Diagnosing issue with resources

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#1662] [Q&A] [answered]
Hello,

I have a project with several modules that use the `resources` directory. When I run the application it does not find the files. The `resources` is at the root of the module as per the documentation. When I execute: 

> /mill -i resolve simulation.resources

I get:
```
[1/1] resolve 
simulation.resources
```
In the directory `/out/simulation/resources.json` I have:

```json
{
    "value": [
        "ref:0a0d0102:/home/hmf/VSCodeProjects/aiops/simulation/resources"
    ],
    "valueHash": 919644686,
    "inputsHash": 132564807
}
```

The command:

> ls -l /home/hmf/VSCodeProjects/aiops/simulation/resources/**

Gives me:

```
-rw-rw-r-- 1 hmf hmf 1562 jan  6 11:26 /home/hmf/VSCodeProjects/aiops/simulation/resources/log4j2.xml
-rw-rw-r-- 1 hmf hmf  807 jan  6 11:26 /home/hmf/VSCodeProjects/aiops/simulation/resources/logback.xml
-rw-rw-r-- 1 hmf hmf 1759 jan  6 11:26 /home/hmf/VSCodeProjects/aiops/simulation/resources/simplelogger.properties

/home/hmf/VSCodeProjects/aiops/simulation/resources/data:
total 900
-rw-rw-r-- 1 hmf hmf 305784 jan 10 17:13 ChartPerformanceBenchmark25Hz_V11.1.1_JDK11_JFX13.csv
-rw-rw-r-- 1 hmf hmf 305784 jan 10 17:13 ChartPerformanceBenchmark25Hz_V11.1.2_JDK11_JFX13.csv
-rw-rw-r-- 1 hmf hmf 305784 jan 10 17:13 ChartPerformanceBenchmark25Hz_V8.1.1_JDK8u112.csv
```
But the code:

>      val resource = classOfChart.getResource("./data/ChartPerformanceBenchmark25Hz_V8.1.1_JDK8u112.csv")

always returns a `null`. I looked for resources in the `compile.dest` directory but did not fund any of these files. 

How can I check that my set-up is correct?

I am using Mill `0.10.0-M5`. 

EDIT: Version `0.9.12` is breaking my build because I am using JavaFX modules.

TIA.

**Comment by lefou** [comment, marked as answer]
Out of memory, I'm not sure, the resource path will ever be valid if it starts with a dot ("./"). Also, if the path isn't absolute, it is relative to the package of the class. So unless `classOfChart` is in the default package, you may want to try an absolute path: ```classOfChart.getResource("/data/ChartPerformanceBenchmark25Hz_V8.1.1_JDK8u112.csv")```, as the resources are put into the top of the archive (or when speaking about packages in the classpath: into the default package).

Resources are not put into `out/<..>/compile.dest`. Instead they are added to the classpath or packaged into the jar, depending whether you use `run` or `jar` target.

**Reply by hmf** [reply]
Thank you. Indeed I had the incorrect path - had to add the package name to the resource path. 

> Resources are not put into out/<..>/compile.dest. Instead they are added to the classpath or packaged into the jar, depending whether you use run or jar target.

Ok. I can use the classpath when running with Mill to check. 

Once again thank you.
